extensions/amp-base-carousel/1.0/base-carousel.js

```diff
@@ -17,43 +17,61 @@ import * as Preact from '../../../src/preact';
 import {Arrow} from './arrow';
 import {Scroller} from './scroller';
 import {toChildArray, useRef, useState} from '../../../src/preact';
+import {useMountEffect} from '../../../src/preact/utils';
 
 /**
- * @param {!JsonObject} props
+ * @param {!BaseCarouselDef.Props} props
  * @return {PreactDef.Renderable}
  */
 export function BaseCarousel(props) {
   const {
-    'arrowPrev': arrowPrev,
-    'arrowNext': arrowNext,
-    'children': children,
-    'defaultSlide': defaultSlide,
-    'loop': loop,
+    arrowPrev,
+    arrowNext,
+    children,
+    defaultSlide,
+    loop,
+    slide,
+    onSlideChange,
+    onMount,
+    setAdvance,
     ...rest
   } = props;
   const childrenArray = toChildArray(children);
   const {length} = childrenArray;
-  const [curSlide, setCurSlide] = useState(defaultSlide || 0);
+  const [curSlide, setCurSlide] = useState(slide ? slide : defaultSlide || 0);
```

Comment:
Can you explain this change? This is the default value, so `slide` shouldn't matter? What is `slide` supposed to do?

Reply:
`slide` and `defaultSlide` are supposed to parallel `value` and `defaultValue` for controlled versus uncontrolled inputs. This particular case is to allow `amp-base-carousel` to mutate the `slide` prop in order to display a different slide. This is helpful for the `goToSlide` action.

I hear you in the next line there is `slideState` which also considers them both. I don't feel strongly about this change and can revert it.

Reply:
Ok, this is because the component can either be controlled or uncontrolled. If I remember correctly, we had a helper from the React Training that would make this a bit easier…

Reply:
I recall the discussion but not the resolution unfortunately 😅  @dvoytenko @samouri any ideas?

Reply:
In controlled mode, there are some expectations:

1. A prop (e.g. `currentSlide`) that controls the parent slide is passed. The carousel can only display that slide.
2. A prop callback (e.g. `onSlideChange`) is called with the new intended slide. The caller updates `currentSlide` prop if it decides to update the carousel. Thus the even handler is critical to the carousel's functioning.
3. Notice that the state is not really needed here.
4. There's no need for public API to inquire the current slide, because the caller already knows it - it's the `currentSlide` prop.

In a non-controlled mode, the expectations are different:

1. A prop (e.g. `defaultSlide`) can control the initially open slide.
2. There's an internal state that tracks the current slide. It goes out of sync with the `defaultSlide`.
3. We'd still likely to have `onSlideChange` - but it's just an event with no expectations attached. It's not critical to the carousel's functioning.
4. We may need an imperative API so that caller could find out the current slide and change it imperatively.

This is something we tried in the prototype, but looking again at this, I'm not sure if the controlled mode is working well here. Although maybe we can find a compromise. But consider this: a user swipes and the carousel is scrolling gradually to the next slide. At some point the controlled `onSlideChange` is called (currently once we are 50% there). Here's what happens:
* If the caller wants to advance to the next slide and updates the prop - the scrolling transition is kind of in the way. If we preserve it - that's a lot of work. If we break it - the effect is very jarring.
* If the caller doesn't want to advance - the user experience is also very jarring - the carousel will jump back.

So, at this point, I'm not sure the controlled mode is such a good idea. I also don't think that opening a carousel to a non-first slide is a useful feature either. It definitely hurts with SSR. And sooner or later we have to look into imperative API anyway.

This is a long way to say that I think we should rip out the `slide` and `defaultSlide` props and controlled mode in general. We can keep `onSlideChange` however.